compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp

```diff
@@ -693,13 +727,119 @@ void distributeReductionBroadcastTranspose(
+    simdToSimtMap.try_emplace(bbArgs[i], bbArgs[i + numOldResults]);
+    if (layoutMap.count(results[i]))
+      layoutMap.try_emplace(results[i + numOldResults],
+                            layoutMap.at(results[i]));
+    simdToSimtMap.try_emplace(results[i], results[i + numOldResults]);
+  }
+
+  rewriter.eraseOp(loop);
+  return;
 }
 
-static void eraseOps(SmallVectorImpl<Operation *> &opsToErase,
-                     IRRewriter &rewriter) {
+void distributeFor(scf::ForOp forOp, DenseMap<Value, Layout> &layoutMap,
+                   DenseMap<Value, Value> &simdToSimtMap, IRRewriter &rewriter,
+                   llvm::SetVector<Operation *> &ops) {
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(forOp);
+
+  SmallVector<Value> newOperands;
+  for (const auto &operand : llvm::enumerate(forOp.getIterOperands())) {
+    APFloat floatValue(0.0);
+    if (matchPattern(operand.value(), m_ConstantFloat(&floatValue))) {
+      if (!layoutMap.count(operand.value())) continue;
+      Layout layout = layoutMap.at(operand.value());
+      Type elementType =
+          operand.value().getType().cast<VectorType>().getElementType();
+      auto vType = VectorType::get(
+          {layout.shape[DimType::Batch0], layout.shape[DimType::Batch1],
+           layout.shape[DimType::VecIdZ] * layout.shape[DimType::VecIdY],
+           layout.shape[DimType::VecIdX]},
+          elementType);
+      Value acc = rewriter.create<arith::ConstantOp>(
+          forOp.getLoc(), DenseElementsAttr::get(vType, floatValue));
+      simdToSimtMap.try_emplace(operand.value(), acc);
+    }
+    if (!simdToSimtMap.count(operand.value())) {
+      continue;
+    }
+    newOperands.push_back(simdToSimtMap.at(operand.value()));
+  }
+  replaceForOpWithNewSignature(rewriter, forOp, newOperands, layoutMap,
+                               simdToSimtMap);
+}
+
+void distributeYield(scf::YieldOp yieldOp, DenseMap<Value, Layout> &layoutMap,
+                     DenseMap<Value, Value> &simdToSimtMap,
+                     IRRewriter &rewriter, llvm::SetVector<Operation *> &ops) {
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(yieldOp);
+
+  // Update yield op with additional operand
+  auto loop = cast<scf::ForOp>(yieldOp->getParentOp());
+  auto yieldOperands = llvm::to_vector<4>(yieldOp.getOperands());
+  for (const auto &operand : llvm::enumerate(yieldOp.getOperands())) {
+    if (!simdToSimtMap.count(operand.value())) continue;
+    // Replace the yield of old value with the for op argument to make it easier
+    // to remove the dead code.
+    yieldOperands[operand.index()] = loop.getIterOperands()[operand.index()];
+    yieldOperands.push_back(simdToSimtMap.at(operand.value()));
+  }
+  rewriter.replaceOpWithNewOp<scf::YieldOp>(yieldOp, yieldOperands);
+}
+
+static void eraseOps(llvm::SetVector<Operation *> &opsToErase,
+                     func::FuncOp funcOp, IRRewriter &rewriter) {
```

Comment:
why do you need the extra arg?

Reply:
It was an artifact from a previous iteration. Will remove.